### ISSUE TITLE
Smallest half-precision float is wrong?

### ISSUE THREAD

**Issue by mrjbq7** [factor/factor#2167]
Looking through the examples on the Wikipedia page:

https://en.wikipedia.org/wiki/Half-precision_floating-point_format#Half_precision_examples

I noticed that the smallest positive subnormal number is wrong:

```factor
USE: math.floats.half

{ 5.960464477539063e-8 } [ 1 bits>half ] unit-test
```

The other examples seem to work...

**Comment by jonenst** [comment]
Looking at the code, it totally wrong for subnormals. You need to convert them to normals. Here's a quick jab at it:
```diff
diff --git a/basis/math/floats/half/half.factor b/basis/math/floats/half/half.factor
index bcb37879d2..4a8ab969c4 100644
--- a/basis/math/floats/half/half.factor
+++ b/basis/math/floats/half/half.factor
@@ -23,8 +23,13 @@ IN: math.floats.half
         0x7fff bitand
         dup zero? [
             dup 0x7c00 >= [ 13 shift 0x7f800000 bitor ] [
+                dup 0x0400 < [
+                dup integer-log2
+                [ nip 103 + 23 shift ] [ 23 swap - shift 0x7fffff bitand ] 2bi bitor
+                ] [
                 13 shift
                 112 23 shift +
+                ] if
             ] if
         ] unless
     ] bi bitor bits>float ;
```
And then you get 
IN: scratchpad 1 bits>half [ . ] [ .h ] bi
5.960464477539062e-08
0x1.0p-24

(basically you need to find the position N of the first set bit in the mantissa, multiply by 2^N and discard the highest bit, and use N as the exponent)

half>bits is also wrong, haven't had time to do it

**Comment by jonenst** [comment]
By the way I didn't look if using integer-logs was OK dependency wise in floats.half

**Comment by mrjbq7** [comment]
Isn't that working on integers already?  So you can just use ``log2``?

Would you be able to fix both words and make a PR?  Or should I try and look at it?

Thank you!

**Comment by jonenst** [comment]
Ah yes I forgot about log2.

I googled around a bit and found http://www.fox-toolkit.org/ftp/fasthalffloatconversion.pdf which is interesting. It says that fixing half>bits is harder.

Not sure when I have time to do a proper PR though..

**Comment by mrjbq7** [comment]
While investigating this, I noticed this weirdness.

@jonenst @erg 

```factor
IN: scratchpad 0x7f802000 dup bits>float float>bits =
f
```

Seems like we turn ``0x7f802000`` into a float, but when it comes back into bits, it's ``0x7fc02000``.  That's a bit... weird!

**Comment by mrjbq7** [comment]
These are NaN... is there a problem casting ``(float)double_value`` with NaN?

**Comment by mrjbq7** [comment]
I think this explains all the failed test cases for my patch which fixes float16.

**Comment by erg** [comment]
We should also add the C code for round-tripping to the vm maybe?

**Comment by mrjbq7** [comment]
Round tripping float16?  Eh, maybe.

It's fine for it to be in Factor, but it shouldn't be buggy.

I think if we figure out why NaN isn't round-tripping only using primitive calls ``float_bits.hpp``, then we commit and can close this issue.

**Comment by erg** [comment]
I'm talking about this issue https://github.com/factor/factor/issues/2083 -- where did we end up with it?

**Comment by mrjbq7** [comment]
I didn't like how Ryu prints stuff that should be simple.

E.g., "123.0" becomes "1.23e2".

**Comment by erg** [comment]
I think the final output is negotiable; the digits are the important part.

**Comment by mrjbq7** [comment]
Fixed (except for NaN) in 6492f1c9cb4e8390228636ac057ae12dd95e14ec.
